File: gui_apps/ch01/layout/layout.cc
// ---------------------------------------------------------------------------
// ch01/layout/main.cc: illustrated use of layouts
// This application will create 4 QLabels and lay them out on a QGridLayout
// which is itself enclosed in a QVBoxLayout
// ---------------------------------------------------------------------------
#include <QApplication>
#include <QDesktopWidget>
#include <QGridLayout>
#include <QLabel>
#include <QString>
#include <QVBoxLayout>

/**
 * @brief centers a widget on the desktop
 * @param
 * 	window - C++ reference the QWidget to center on desktop
 * @returns
 * 	none
 */
void showCenteredOnDesktop(QWidget &window)
{
   window.show();
   QRect screenSize = QApplication::desktop()->screenGeometry();
   int x = (screenSize.width() - window.width()) / 2;
   int y = (screenSize.height() - window.height()) / 2;
   window.move(x, y);
}

int main(int argc, char **argv)
{
   QGuiApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
   QApplication app(argc, argv);
   QFont font("SF UI Text", 10);
   QApplication::setFont(font);

   // create & show the GUI
   QWidget window;
   window.setWindowTitle("Qt Layouts Example");
   QVBoxLayout *layout = new QVBoxLayout;

   QString messages[] = {"Hello", "World", "Welcome", "to the awesome Qt Framework"};
   QGridLayout *glayout = new QGridLayout();
   glayout->setHorizontalSpacing(5);
   glayout->setVerticalSpacing(5);
   // layout the first 3 strings horizontally in the first row
   glayout->addWidget(new QLabel(messages[0]), 0, 0);
   glayout->addWidget(new QLabel(messages[1]), 0, 1);
   glayout->addWidget(new QLabel(messages[2]), 0, 2);
   // the 4th string is layed out in the second row, spanning all three cols
   QLabel *lbl = new QLabel(messages[3]);
   lbl->setAlignment(Qt::AlignCenter);
   glayout->addWidget(lbl, 1, 0, 1, 3);
   layout->addLayout(glayout);
   window.setLayout(layout);
   showCenteredOnDesktop(window);

   return app.exec();
}

File: gui_apps/ch01/layout/layout.py
# layout.py: illustrates layouts in PyQt
import sys
import pathlib
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

USING_PYQT6 = True if PYQT_VERSION_STR.startswith('6') else False

p = pathlib.Path(__file__)
print(f"Current file's path is {str(p)}")
p.parents[2]
print(f"p.parents[2] = {str(p.parents[2])}")
sys.path.append(str(p.parents[2]))
import globalvars
globalvars.USING_PYQT6 = USING_PYQT6
import utils


class LayoutWidget(utils.CenteredOnDesktopWidget):
    def __init__(self, *args, **kwargs):
        super(utils.CenteredOnDesktopWidget, self).__init__(*args, **kwargs)
        self.setWindowTitle("Qt Layouts Example")

        messages = ["PyQt5", "makes", "creating",
                    "GUI apps really easy!"]

        layout = QVBoxLayout()
        lblLayout = QHBoxLayout()
        glayout = QGridLayout()
        glayout.horizontalSpacing = 5
        glayout.verticalSpacing = 5

        for i, msg in enumerate(messages):
            #lbl = QLabel(msg)
            btn = QPushButton(msg)
            if i == len(messages) - 1:
                # lbl.setAlignment(Qt.AlignCenter)
                glayout.addWidget(btn, 1, 0, 1, i)
            else:
                glayout.addWidget(btn, 0, i)
        lbl = QLabel("")  # will display
        layout.addLayout(glayout)
        layout.addLayout(lblLayout)
        self.setLayout(layout)


def main():
    app = QApplication(sys.argv)
    #font = QFont(app.font().family(), 11)
    # app.setFont(font)
    app.setFont(QApplication.font("QMenu"))

    w = LayoutWidget()
    w.show()
    return app.exec()


if __name__ == '__main__':
    main()
